commit 59c2fe1af1bc2b3107140668c9237aef9d3ad392
Author: Rafael Espindola <rafael.espindola@gmail.com>
Date:   Thu Jan 16 04:50:58 2014 +0000

    Prevent calls to __jit_debug_register_code from being optimized out.
    
    Patch by Andrew MacPherson. I just tweaked the comment.
    
    git-svn-id: https://llvm.org/svn/llvm-project/llvm/trunk@199350 91177308-0d34-0410-b5e6-96231b3b80d8
diff -u -r llvm-3.4.2.src/lib/ExecutionEngine/RuntimeDyld/GDBRegistrar.cpp llvm-3.4.2.src.fix/lib/ExecutionEngine/RuntimeDyld/GDBRegistrar.cpp
--- llvm-3.4.2.src/lib/ExecutionEngine/RuntimeDyld/GDBRegistrar.cpp	2012-12-24 01:42:27.000000000 -0800
+++ llvm-3.4.2.src.fix/lib/ExecutionEngine/RuntimeDyld/GDBRegistrar.cpp	2015-03-17 20:19:19.870787031 -0700
@@ -47,7 +47,11 @@
   struct jit_descriptor __jit_debug_descriptor = { 1, 0, 0, 0 };
 
   // Debuggers puts a breakpoint in this function.
-  LLVM_ATTRIBUTE_NOINLINE void __jit_debug_register_code() { }
+  LLVM_ATTRIBUTE_NOINLINE void __jit_debug_register_code() {
+    // The noinline and the asm prevent calls to this function from being
+    // optimized out.
+    asm volatile("":::"memory");
+  }
 
 }
 
